src/main/clojure/clara/rules/compiler.clj

```diff
@@ -235,15 +235,21 @@
   "Compiles a sequence of constraints into a structure that can be evaluated.
 
    Callers may also pass a collection of equality-only-variables, which instructs
-   this function to only do an equality check on them rather than create a unification binding."
+   this function to only do an equality check on them rather than create a unification binding.
+
+   The returned expression evaluates to a two item vector, [bindings error], which contains
+   the evaluated bindings in the first argument and a map with information on any exception
+   that might have been thrown by evaluating a constraint as the second vector entry."
```

Comment:
That makes sense, but can we get the exact throwing constraint this way? Sequences of constraints are often compiled together by `compile-constraints`. 

We can get `?__bindings__` by capturing exceptions in `compile-condition` and `compile-join-filter` functions. We can do similarly for constraint expressions if we store the current expression in an atom just like bindings, but I'm not sure whether that would outperform the original vector-based approach. I imagine that atoms themselves have some overhead for thread synchronization. Also, we would no longer be able to distinguish between a constraint expression throwing and Clara throwing. This distinction can have an impact on users' confidence in the system.

Beyond setting up some benchmarks, I'm not sure how to make the performance call.